typed-rng/src/lib.rs

```diff
@@ -0,0 +1,234 @@
+//! generated with property-based testing, with e.g. `Arbitrary` or proptest's
+//! `Strategy`.
+//!
+//! But with property-based tests, the goal is for small changes in the random
+//! input to result in _small_ changes in the output. High-quality libraries
+//! like proptest also focus on shrinking.
+//!
+//! With `Generatable`, the goal is for small changes in the random input (the
+//! seed used to initialize the RNG) to result in _large_ changes in output.
+//! (However, it is possible to delegate the bulk of the operation to a value
+//! generator from a PBT framework). There is also no need for shrinking.
+//!
+//! Overall, this means that `Generatable` can be used in both testing and
+//! production.
+//!
+//! ## Other functionality
+//!
+//! This crate also provides two additional convenience functions:
+//!
+//! - [`from_seed`], which generates a new RNG from a seed.
+//! - [`from_parent_and_seed`], which generates a new RNG from a parent RNG and
+//!   a seed.
+//!
+//! Both these methods are short, but more ergonomic and less prone to misuse
+//! than using the underlying libraries directly.
+
+use std::{fmt, hash::Hash, marker::PhantomData};
+
+use rand::rngs::StdRng;
+use rand_core::{RngCore, SeedableRng};
+use uuid::Uuid;
+
+/// Returns a new RNG generated only from the given seeds `seed` and `extra`.
+/// `seed` may be passed down from another caller, e.g. a test, and `extra`
+/// should be a fixed value specific to the callsite.
+///
+/// This takes two hashable arguments rather than one, because when one is
+/// passing down a seed, it is all too easy to not include any extra
+/// information. That may result in multiple different RNGs generating the same
+/// random values. So we expect that callers will also provide bytes of their
+/// own, specific to the callsite, and gently guide them towards doing the
+/// right thing.
+pub fn from_seed<R, H, H2>(seed: H, extra: H2) -> R
+where
+    R: SeedableRng,
+    H: Hash,
+    H2: Hash,
+{
+    // XXX: is Hash really the right thing to use here? That's what
+    // rand_seeder uses, but something like https://docs.rs/stable-hash may
+    // be more correct.
+
+    let mut seeder = rand_seeder::Seeder::from((seed, extra));
+    seeder.make_rng::<R>()
+}
+
+/// Generates a new RNG from a parent RNG and a hashable seed.
+pub fn from_parent_and_seed<R, R2, H>(parent_rng: &mut R2, seed: H) -> R
+where
+    R: SeedableRng,
+    R2: RngCore,
+    H: Hash,
+{
+    let rng_seed = parent_rng.next_u64();
+
+    let mut seeder = rand_seeder::Seeder::from((rng_seed, seed));
+    seeder.make_rng::<R>()
+}
+
+/// An RNG that can be used to generate values of a single type.
+///
+/// This is a convenience wrapper around a random number generator that
+/// generates values of a particular type. It works against any type that
+/// implements [`Generatable`], and any RNG that implements [`RngCore`].
+pub struct TypedRng<T, R> {
+    rng: R,
+    _marker: PhantomData<T>,
```

Comment:
Tiny nit - should this be `PhantomData<fn() -> T>` or similar to avoid `TypedRng` losing `Send` etc if `T` isn't `Send`?

Reply:
TIL, thanks! https://doc.rust-lang.org/nomicon/phantom-data.html#table-of-phantomdata-patterns